File: px4_offboard/capture_video.py
#!/usr/bin/env python3
import rclpy 
from rclpy.node import Node 

from sensor_msgs.msg import Image 
from cv_bridge import CvBridge 
import cv2
import os

class CaptureCvVideo(Node):
    def __init__(self):
        super().__init__("cv_video_capture")

        self.subscription = self.create_subscription(
            Image,
            '/camera',
            self.capture_cv,
            10
        )
        self.br = CvBridge()
        self.save_dir = "/paraceptor/cv_capture_data"
        self.save_every_n_frame = 5
        self.frame_count = 0
        self.image_counter = 0

    def capture_cv(self, data):
        self.frame_count += 1

        current_frame = self.br.imgmsg_to_cv2(data, desired_encoding="bgr8")
       
        if self.frame_count % self.save_every_n_frame == 0:
            self.image_counter += 1
            filename = os.path.join(self.save_dir, f"img_{self.image_counter}.png")

            try:
                cv2.imwrite(filename, current_frame)
            except Exception as e:
                self.get_logger().error(f"Failed to save image {filename}: {e}")

        cv2.imshow("camera feed", current_frame)

def main(args=None):
    rclpy.init(args=args)
    image_subscriber = CaptureCvVideo()
    rclpy.spin(image_subscriber)
    image_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: px4_offboard/capture_video.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include <cv_bridge/cv_bridge.h>
#include <opencv2/highgui.hpp>
#include <opencv2/opencv.hpp>

using namespace cv;

class CaptureCVData : public rclcpp::Node
{
public: 
 CaptureCVData() : Node("cv_data_subscription")
 {

    subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
        "/camera", 
        10,
        [this](const sensor_msgs::msg::Image::SharedPtr msg){
            this->screenshot_data(msg);
        });
 }
    void screenshot_data(const sensor_msgs::msg::Image::SharedPtr image)
    {
        cv_bridge::CvImagePtr cv_ptr;
        try
        {
            cv_ptr = cv_bridge::toCvCopy(image, sensor_msgs::image_encodings::BGR8);
        }
        catch (cv_bridge::Exception& e)
        {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }
        cv::Mat frame = cv_ptr->image;

        cv::imshow("camera_feed", frame);
        cv::waitKey(1);
    }
private:
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
    
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<CaptureCVData>());
    rclcpp::shutdown();
    return 0;

}